arrow-fx-coroutines/README.MD

```diff
@@ -1,561 +1,561 @@
 
-becomes
+## Kotlin Std Coroutines
+### CoroutineContext
 
-```kotlin:ank
-suspend fun getThreadName(): String =
-  Thread.currentThread().name
+CoroutineContext is a immutable key-value map, where the keys are the types and the values are the instances of the type. Meaning that for every type, there can only be a single instance entry.
+This is useful to for example ensure there is only a single `ContinuationInterceptor`, or a single `Job` in the case of KotlinX Coroutines or a single `SuspendConnection` in the case of Arrow Fx Coroutines.
 
-suspend fun main(): Unit {
-  val t1 = evalOn(Computation) { getThreadName() }
-  val t2 = evalOn(BlockingIO) { getThreadName() }
-  val t3 = evalOn(UI) { getThreadName() }
-  println("$t1 ~> $t2 ~> $t3")
-}
-```
+A `suspend` function/program **always** has a `CoroutineContext`, this restriction is made in the `Continuation` interface.
 
-As you can see we can directly apply `evalOn` to our `suspend fun`, this eliminates the need to wrap explicitly in `IO`.
-Both programs are equivalent in semantics and guarantees, the later will however perform better since it eliminates the wrapping `IO` requires.
-This is possible without losing cancellation support, as explained in the detail here [anchor cancellation].
+This type-instance relationship is achieved by two interfaces `CoroutineContext.Element` and `CoroutineContext.Key<E: CoroutineContext.Element>`. Where `CoroutineContext.Element` has a `Key<*>`, which needs to point to a `Key` for it's concrete type. 
 
-## IO<Either<E, A>> vs suspend () -> Either<E, A>
+A simple example:
 
-When writing functional code style we often want to express our domain errors as clearly as possible, a popular pattern is to return `Either<DomainError, SuccessValue>`.
-Let's assume following domain, and compare two snippets one using `IO<Either<E, A>>` and another `suspend () -> Either<E, A>`.
+```kotlin
+class CoroutineName(val name: String): CoroutineContext.Element {
+  companion object Key : CoroutineContext.Key<CoroutineName>
+  override val key: CoroutineContext.Key<*> = Key
+}
+```
 
-```kotlin:ank
-inline class Id(val id: Long)
-object PersistenceError
+### Continuation
 
-data class User(val email: String, val name: String)
-data class ProcessedUser(val id: Id, val email: String, val name: String)
+```kotlin
+interface Continuation<A> {
+	val ctx: CoroutineContext
+	fun resume(result: Result<A>)
+}
+```
 
-suspend fun fetchUser(): Either<PersistenceError, User> =
-    Right(User("simon@arrow-kt.io", "Simon"))
+As mentioned above, a `suspend` programs can return immediately or `suspend` and return it's value at aa later point in time. Thus this can be modeled with a `Contination<A>` where both the immediate result or a suspended result is always returned to `resume`, this ensures that the result is always consumed in the same manner no matter how the `suspend` program behaves.
```

Comment:
```suggestion
As mentioned above, a `suspend` programs can return immediately or `suspend` and return it's value at a later point in time. Thus this can be modeled with a `Contination<A>` where both the immediate result or a suspended result is always returned to `resume`, this ensures that the result is always consumed in the same manner no matter how the `suspend` program behaves.
```